drivers/adc/ad7799/ad7799.c

```diff
@@ -0,0 +1,444 @@
+ */
+int32_t ad7799_reset(struct ad7799_dev *device)
+{
+	uint8_t reset_data[4] = {
+		AD7799_RESET_DATA,
+		AD7799_RESET_DATA,
+		AD7799_RESET_DATA,
+		AD7799_RESET_DATA
+	};
+
+	return spi_write_and_read(device->spi_desc, reset_data, 4);
+}
+
+/**
+ * @brief Set the device mode.
+ * @param device - The device structure.
+ * @param mode - The device mode.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t ad7799_set_mode(struct ad7799_dev *device, uint8_t mode)
+{
+	int32_t ret;
+	uint32_t reg_data = 0;
+
+	ret = ad7799_read(device, AD7799_REG_MODE, &reg_data);
+	if (ret)
+		return FAILURE;
+
+	reg_data &= ~AD7799_MODE_SEL(AD7799_REG_MASK);
+	reg_data |= AD7799_MODE_SEL(mode);
+
+	ret = ad7799_write(device, AD7799_REG_MODE, reg_data);
+	if (ret)
+		return FAILURE;
+
+	ret = ad7799_dev_ready(device);
+	if (ret)
+		return FAILURE;
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Select the ADC channel.
+ * @param device - The device structure.
+ * @param  ch - The channel number.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t ad7799_set_channel(struct ad7799_dev *device, uint8_t ch)
+{
+	int32_t ret;
+	uint32_t reg_data = 0;
+
+	ret = ad7799_read(device, AD7799_REG_CONF, &reg_data);
+	if (ret)
+		return FAILURE;
+
+	reg_data &= ~AD7799_CONF_CHAN(AD7799_REG_MASK);
+	reg_data |= AD7799_CONF_CHAN(ch);
+
+	return ad7799_write(device, AD7799_REG_CONF, reg_data);
+}
+
+/**
+ * @brief Read specific ADC channel.
+ * @param device - The device structure.
+ * @param ch - The ADC channel.
+ * @param reg_data - The content of the data register.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t ad7799_get_channel(struct ad7799_dev *device, uint8_t ch,
+			   uint32_t *reg_data)
+{
+	int32_t ret;
+
+	ret = ad7799_set_channel(device, ch);
+	if (ret)
+		return FAILURE;
+
+	ret = ad7799_set_mode(device, AD7799_MODE_SINGLE);
+	if (ret)
+		return FAILURE;
+
+	ret = ad7799_dev_ready(device);
+	if (ret)
+		return FAILURE;
+
+	ret = ad7799_read(device, AD7799_REG_DATA, reg_data);
+	if (ret)
+		return FAILURE;
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Set the ADC gain.
+ * @param device - The device structure.
+ * @param  gain - the channel number.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t ad7799_set_gain(struct ad7799_dev *device, uint8_t gain)
+{
+	int32_t ret;
+	uint32_t reg_data = 0;
+
+	ret = ad7799_read(device, AD7799_REG_CONF, &reg_data);
+	if (ret)
+		return FAILURE;
+
+	reg_data &= ~AD7799_CONF_GAIN(AD7799_REG_MASK);
+	reg_data |= AD7799_CONF_GAIN(gain);
+
+	return ad7799_write(device, AD7799_REG_CONF, reg_data);
+}
+
+/**
+ * @brief Get the ADC gain
+ * @param device - The device structure.
+ * @param  gain - the gain value from the register.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t ad7799_get_gain(struct ad7799_dev *device, uint8_t *gain)
```

Comment:
You could drop this function in favour of a `device->gain` member that works as a cached setting.

Reply:
This function was created only for test purposes (when testing the driver in conjunction with IIO). 
I thought I can leave it since it was already implemented.

Otherwise, if required, I can remove it.

In my opinion, adding `gain` member to the structure implies adding also members for other driver specific parameters (polarity, channel selection, ref detect, etc.)

Reply:
I agree, adding `gain` would imply adding other structure members too (that's what we do in other drivers).
If you don't wanna readback these values I guess deleting this function would be fine too.